MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -397,6 +682,47 @@ DumpBitmap (
   DEBUG ((DEBUG_INFO, "\n"));
 }
 
+/**
+  Debug dumps the input bitmap
+
+  @param[in] Bitmap Pointer to the start of the bitmap
+  @param[in] Count  Number of bitmap entries
+**/
+STATIC
+VOID
+DumpMemoryProtectionSpecialRegions (
+  VOID
+  )
+{
+  MEMORY_PROTECTION_SPECIAL_REGION  *SpecialRegions = NULL;
+  UINTN                             Count           = 0;
+  UINTN                             Index           = 0;
+  EFI_STATUS                        Status;
+
+  Status = GetSpecialRegions (&SpecialRegions, &Count);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_WARN,
+      "Failed to fetch the memory protection special regions! Status: %r\n",
+      Status
+      ));
+    return;
+  }
+
+  for ( ; Index < Count; Index++) {
+    DEBUG ((
+      DEBUG_INFO,
+      "Memory Protection Special Region: 0x%llx - 0x%llx. Attributes: 0x%llx\n",
+      SpecialRegions[Index].Start,
+      SpecialRegions[Index].Start + SpecialRegions[Index].Length,
+      SpecialRegions[Index].EfiAttributes
+      ));
+  }
+

```

Comment:
is there a memory leak here with GetSpecialRegions allocating memory and no free?

Reply:
I think this is related to your request for a comment in the header stating that the allocated array should be freed by the caller. Is that right?

Reply:
yes.

DumpMemoryProtectionSpecialRegions()
calls GetSpecialRegions () then goes through th results and print info

GetSpecialRegions() allocates the buffer it returns:

SpecialRegions = AllocatePool (sizeof (MEMORY_PROTECTION_SPECIAL_REGION) * mSpecialMemoryRegionsPrivate.Count);

DumpMemoryProtectionSpecialRegions() isn't freeing that buffer

Reply:
DumpMemoryProtectionSpecialRegions seems to be an internal debug function, would it be better to just go through the special region list directly and not need to call allocate?

Reply:
Ah I misunderstood where this comment was in the code - this has been fixed :)

Reply:
I went ahead and updated this dump to just use the global list instead of the GetSpecialRegions() interface to save an allocation. Thanks for the suggestion!